docs/security/remediation/validate_ledger.py

```
@@ -0,0 +1,1023 @@
+        require(root_by_source[duplicate] == root_by_source[primary], f"{duplicate}: exact duplicate must inherit {primary} root")
+    for start in EXACT_DUPLICATES:
+        seen: set[str] = set()
+        current = start
+        while current in EXACT_DUPLICATES:
+            require(current not in seen, f"{start}: cyclic exact-duplicate mapping")
+            seen.add(current)
+            current = EXACT_DUPLICATES[current]
+
+
+def git_show_json(base_ref: str, relative_path: str) -> Any | None:
+    result = subprocess.run(
```

Ruff (0.16.1)

[error] `subprocess` call: check for execution of untrusted input

(S603)